crds/main.go

```diff
@@ -0,0 +1,374 @@
+		Language: map[string]json.RawMessage{
+			"nodejs": []byte("{}"),
+		},
+	}
+
+	// Convert the PackageSpec into a Package.
+	pkg, err := schema.ImportSpec(pkgSpec, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// Generate all the code for the package.
+	files, err := GeneratePackage(tool, pkg, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// Extract the relevant generated code.
+	// Here, I'm just returning the content of the in-memory generated `types/input.ts` file,
+	// which has generated interfaces of the "input shape". Not sure if we want the "output shape"
+	// instead (or in addition to) that are in `types/output.ts`, or if we want something entirely different.
+	if file, ok := files["types/input.ts"]; ok {
+		return string(file), nil
+	}
+
+	// Note: The generated code isn't quite right for what we need for the CRD code gen. For example,
+	// There are some comments at the top (with "WARNING") about being generated code that we probably
+	// don't want to include. There are some imports that don't make sense outside of package code
+	// generation, and I'm not sure we want to use TypeScript namespaces.
+
+	// But it does give a sense for what's possible calling into existing code.
+
+	// I think you will need to refactor some of the existing code in
+	// https://github.com/pulumi/pulumi/blob/master/pkg/codegen/nodejs/gen.go and expose a
+	// public function we can call that will generate the right code. Perhaps leveraging/refactoring
+	// some existing code we have and generalizing it so we can use it for this purpose.
+
+	// Maybe something like generalizing, refactoring, and making public the existing `genPlainType` function?
+	// https://github.com/pulumi/pulumi/blob/bb358c4d217334ad654f153cc155f84a4fb1933b/pkg/codegen/nodejs/gen.go#L243
+
+	// And then we'll want to do something similar for each of the other languages (C#, Go, Python).
+
+	return "", nil
+}
+
+const anyTypeRef = "pulumi.json#/Any"
+
+// Returns the designated "any" TypeSpec
+func anyTypeSpec() pschema.TypeSpec {
+	return pschema.TypeSpec{
+		Ref: anyTypeRef,
+	}
+}
+
+// Returns true if the given TypeSpec is of type any; returns false otherwise
+func isAnyType(typeSpec pschema.TypeSpec) bool {
+	return typeSpec.Ref == anyTypeRef
+}
+
+// Converts a []string to []interface{}
+func genericizeStringSlice(stringSlice []string) interface{} {
+	genericSlice := make([]interface{}, len(stringSlice))
+	for i, v := range stringSlice {
+		genericSlice[i] = v
+	}
+	return genericSlice
+}
+
+// Given a list of schemas, returns a single schema representing their
+// intersection type. Combines each of their properties, and descriptions into a single
+// schema. If combineRequired is true, then required properties are also
+// combined into the returned schema; otherwise all properties in the returned
+// schema are optional.
+func combineSchemas(combineRequired bool, schemas ...map[string]interface{}) map[string]interface{} {
```

Comment:
It would be good to add a unit test to make sure this is working as expected. This logic is hard to verify by inspection.